node/core/dispute-coordinator/src/scraping/mod.rs

```diff
@@ -92,24 +92,36 @@ impl Inclusions {
 		block_hash: Hash,
 	) {
 		if let Some(blocks_including) = self.inclusions_inner.get_mut(&candidate_hash) {
-			let first_lower_entry = blocks_including.partition_point(|(number_at_idx, _)| *number_at_idx >= block_number);
-			blocks_including.insert(first_lower_entry, (block_number, block_hash));
+			if let Some(blocks_at_height) = blocks_including.get_mut(&block_number) {
+				blocks_at_height.push(block_hash);
+			} else {
+				blocks_including.insert(block_number, Vec::from([block_hash]));
+			}
 		} else {
+			let mut blocks_including: BTreeMap<BlockNumber, Vec<Hash>> = BTreeMap::new();
+			blocks_including.insert(block_number, Vec::from([block_hash]));
 			self.inclusions_inner
-				.insert(candidate_hash, Vec::from([(block_number, block_hash)]));
+				.insert(candidate_hash, blocks_including);
 		}
 	}
 
 	pub fn remove_up_to_height(&mut self, height: &BlockNumber) {
-		for including_blocks in self.inclusions_inner.values_mut() {
-			let first_lower_entry = including_blocks.partition_point(|(number_at_idx, _)| *number_at_idx >= *height);
-			including_blocks.drain(first_lower_entry..);
+		for blocks_including in self.inclusions_inner.values_mut() {
+			*blocks_including = blocks_including.split_off(height);
 		}
-		self.inclusions_inner.retain(|_, including_blocks| including_blocks.len() > 0);
+		self.inclusions_inner.retain(|_, including_blocks| including_blocks.keys().len() > 0);
 	}
 
-	pub fn get(&mut self, candidate: &CandidateHash) -> Option<&Vec<(BlockNumber, Hash)>> {
-		self.inclusions_inner.get(candidate)
+	pub fn get(&mut self, candidate: &CandidateHash) -> Vec<(BlockNumber, Hash)> {
```

Comment:
Sorry, I was a bit too quick with my last review and missed that this is still nested and we still need to traverse all candidates on each block getting finalized. Traversing potentially thousands of entries every 6 seconds (if there is a slight finality lag), might not be a biggie, but it is not pretty either. Your initial idea of piggybacking on `RefCountedCandidates` had one good property: Making use of the top-level `BTreeMap<BlockNumber, HashSet<CandidateHash>` in `ScrapedCandidates` could avoid this. We could for example make [`ScrapedCandidates::remove_up_to_height`](https://github.com/paritytech/polkadot/blob/9f6f2ab7df271823dc1f09b187908da1e07beb86/node/core/dispute-coordinator/src/scraping/candidates.rs#L105) return the cleaned up `CandidateHash`es and use those of [`included_candidates`](https://github.com/paritytech/polkadot/blob/992f2405a1a08053cca0f506bcbe39db56adcbe7/node/core/dispute-coordinator/src/scraping/mod.rs#L148) for cleaning up `Inclusions`.

Reply:
Added back the piggybacking 👍